### PR TITLE
Fix Metal tests [WIP]

### PR DESCRIPTION
**Pull request by rezaarezvan**
I've managed to track down why and fix the half-to-int tests, but now remains the conv tests.

After that, all metal tests should pass!

### CODE REVIEW

File: tinygrad/codegen/linearizer.py [hunk 1]
```diff
@@ -566,7 +566,7 @@ def hand_coded_optimizations(self):
     for splits in [4]:
       if self.upcasted == 0 and len(self.full_unupcasted_shape) > 0 and self.full_unupcasted_shape[-1] % splits == 0:
         self.shift_to(len(self.full_unupcasted_shape)-1, splits, insert_before=len(self.full_unupcasted_shape))
-        self.upcast()
+        if self.bufs[0].dtype.name.startswith('image') and self.first_reduce > 0 and self.first_reduce < (self.shape_len-self.upcasted): self.upcast()
```
Comment (geohot) [inline tinygrad/codegen/linearizer.py:569]: Is this a fix or a mitigation?

Comment (rezaarezvan) [reply tinygrad/codegen/linearizer.py:569]: This is a fix according to my testings

Comment (geohot) [reply tinygrad/codegen/linearizer.py:569]: It's not. What's broken that this fixes it?

Comment (rezaarezvan) [reply tinygrad/codegen/linearizer.py:569]: > It's not. What's broken that this fixes it?

From what I've found, the upcast is what breaks these tests. But yeah, to be fair I haven't looked deeper into why it breaks. I'll should look into that

**Pull request closed without merging**